pkg/monitoring/vmistats/collector.go

```diff
@@ -171,16 +175,159 @@ func checkNonEvictableVMAndSetMetric(vmi *k6tv1.VirtualMachineInstance) float64
 	return setVal
 }
 
-func updateVMIEvictionBlocker(vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+func updateVMIMetrics(vmis []*k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
 	for _, vmi := range vmis {
-		mv, err := prometheus.NewConstMetric(
-			vmiEvictionBlockerDesc, prometheus.GaugeValue,
-			checkNonEvictableVMAndSetMetric(vmi),
-			vmi.Status.NodeName, vmi.Namespace, vmi.Name,
-		)
-		if err != nil {
-			continue
+		updateVMIEvictionBlocker(vmi, ch)
+	}
+}
+
+func updateVMIEvictionBlocker(vmi *k6tv1.VirtualMachineInstance, ch chan<- prometheus.Metric) {
+	mv, err := prometheus.NewConstMetric(
+		vmiEvictionBlockerDesc, prometheus.GaugeValue,
+		checkNonEvictableVMAndSetMetric(vmi),
+		vmi.Status.NodeName, vmi.Namespace, vmi.Name,
+	)
+	if err != nil {
+		return
+	}
+	ch <- mv
+
+}
+
+func getTransitionTimeSeconds(fromCreation bool, oldVMI *k6tv1.VirtualMachineInstance, newVMI *k6tv1.VirtualMachineInstance) (float64, error) {
+
+	var oldTime *metav1.Time
+	var newTime *metav1.Time
+
+	if fromCreation || oldVMI == nil || (oldVMI.Status.Phase == k6tv1.VmPhaseUnset) {
+		oldTime = newVMI.CreationTimestamp.DeepCopy()
+	}
+
+	for _, transitionTimestamp := range newVMI.Status.PhaseTransitionTimestamps {
```

Comment:
We can miss edge triggered events like `Scheduled` and that can misrepresent the transition from `Scheduling -> Running`.  It can also misrepresent `Scheduled`, if the majority of VMs were quick and one was slow so the metrics only pickup the slow `Scheduled`.  What do you think about `Scheduled` being merged into `Scheduling -> Running`?

Reply:
> We can miss edge triggered events like Scheduled and that can misrepresent the transition from Scheduling -> Running

I'm not sure that's possible. 

We're not updating the histogram based on a scrape interval. We're actually watching all VMI's and updating the histogram client side from the informer.

So the histogram gets continually updated independently of the scrape interval, and the scrape interval is getting an update of all the new bucket counts, totals, sums, etc... At least that's my understanding

Reply:
>We're actually watching all VMI's and updating the histogram client side from the informer.
>
That's what I'm referring to.  

>We can miss edge triggered events like Scheduled
>
Went back to look at the reconcile code..., `Scheduled` isn't actually an edge triggered event like I thought because of the `-ed` postfix (other examples: Failed, Succeeded, Deleted).  The phase 'Scheduling' is when the virt-controller is waiting for the virt-launcher pod to be ready.  And `Scheduled` is actually a level triggered event where the virt-handler has taken over for reconciling the VMI.  `Scheduling` + `Scheduled` is really `Scheduling`, but I see the value in recording the transition.

The reason why I thought `Scheduled` phase would sometimes be missed is because I thought it was a race between libvirt noticing guest activity and the handler processing that the pod is Running. 

We won't miss any phases.